include/device_queue.h

```diff
@@ -59,9 +65,170 @@ namespace detail {
 		std::unique_ptr<cl::sycl::queue> sycl_queue;
 		bool device_profiling_enabled = false;
 
-		cl::sycl::device pick_device(const config& cfg, cl::sycl::device* user_device) const;
 		void handle_async_exceptions(cl::sycl::exception_list el) const;
 	};
 
+	// Try to find a platform that can provide a unique device for each node using a device selector.
+	template <typename DeviceT, typename PlatformT, typename SelectorT>
+	bool try_find_device_per_node(
+	    std::string& how_selected, DeviceT& device, const std::vector<PlatformT>& platforms, const host_config& host_cfg, SelectorT selector) {
+		std::vector<std::tuple<DeviceT, size_t>> devices_with_platform_idx;
+		for(size_t i = 0; i < platforms.size(); ++i) {
+			auto&& platform = platforms[i];
+			if(platform.get_devices().size() >= host_cfg.node_count) {
+				for(auto device : platform.get_devices()) {
+					if(selector(device) == -1) { continue; }
+					devices_with_platform_idx.emplace_back(device, i);
+				}
+			}
+		}
+
+		std::stable_sort(devices_with_platform_idx.begin(), devices_with_platform_idx.end(),
+		    [selector](const auto& a, const auto& b) { return selector(std::get<0>(a)) > selector(std::get<0>(b)); });
+		bool same_platform = true;
+		bool same_device_type = true;
+		if(devices_with_platform_idx.size() >= host_cfg.node_count) {
+			auto [device_from_platform, idx] = devices_with_platform_idx[0];
+			const auto platform = device_from_platform.get_platform();
+			const auto device_type = device_from_platform.template get_info<sycl::info::device::device_type>();
+
+			for(size_t i = 1; i < host_cfg.node_count; ++i) {
+				auto [device_from_platform, idx] = devices_with_platform_idx[i];
+				if(device_from_platform.get_platform() != platform) { same_platform = false; }
+				if(device_from_platform.template get_info<sycl::info::device::device_type>() != device_type) { same_device_type = false; }
+			}
+
+			if(!same_platform) { CELERITY_WARN("Selected devices do not belong to the same platform"); }
```

Comment:
I feel like having one warning for both would be sufficient here. We print the platform / device name anyways. So maybe `"Selected devices are of different type and/or do not belong to the same platform"`.

Reply:
Thank you for the comments. Yes, it makes sense.